src/TempDirectory.cc

```diff
@@ -0,0 +1,223 @@
+    const std::string &_parentPath)
+{
+  std::string parentPath(_parentPath);
+  std::string templatePath = _baseName + "XXXXXX";
+
+  std::string fullTemplateStr = joinPaths(parentPath, templatePath);
+  if (!createDirectories(parentPath)) {
+    std::error_code ec{errno, std::system_category()};
+    errno = 0;
+    throw std::system_error(ec, "could not create the parent directory");
+  }
+
+#ifdef _WIN32
+  errno_t errcode = _mktemp_s(&fullTemplateStr[0], fullTemplateStr.size() + 1);
+  if (errcode) {
+    std::error_code ec(static_cast<int>(errcode), std::system_category());
+    throw std::system_error(ec,
+        "could not format the temp directory name template");
+  }
+  const std::string finalPath{fullTemplateStr};
+  if (!createDirectories(finalPath)) {
+    std::error_code ec(static_cast<int>(GetLastError()),
+        std::system_category());
+    throw std::system_error(ec, "could not create the temp directory");
+  }
+#else
+  const char * dirName = mkdtemp(&fullTemplateStr[0]);
+  if (dirName == nullptr) {
+    std::error_code ec{errno, std::system_category()};
+    errno = 0;
+    throw std::system_error(ec,
+        "could not format or create the temp directory");
+  }
+  const std::string finalPath{dirName};
+#endif
+
+  return finalPath;
+}
+
+/////////////////////////////////////////////////
+std::string ignition::common::createTempDirectory(
+    const std::string &_baseName,
+    const std::string &_parentPath,
+    const FilesystemWarningOp _warningOp)
+{
+  std::string ret;
+  try {
+    ret = ::createTempDirectory(_baseName, _parentPath);
+  } catch (const std::system_error &ex) {
+    ret = "";
+    if(FSWO_LOG_WARNINGS == _warningOp)
+    {
+      ignwarn << "Failed to create temp directory: " << ex.what() << "\n";
+    }
+  }
+  return ret;
+}
+
+class ignition::common::TempDirectoryPrivate
+{
+  /// \brief Current working directory before creation of temporary dir.
+  public: std::string oldPath {""};
+
+  /// \brief Path of the temporary directory
+  public: std::string path {""};
+
+  /// \brief True if the temporary directory exists
+  public: bool isValid {false};
+
+  /// \brief True if the temporary directory should be cleaned up from
+  /// disk when the object goes out of scope.
+  public: bool doCleanup {true};
+};
+
+/////////////////////////////////////////////////
+TempDirectory::TempDirectory(const std::string &_prefix,
+                             const std::string &_subDir,
+                             bool _cleanup):
+  dataPtr(std::make_unique<TempDirectoryPrivate>())
+{
+
+  this->dataPtr->oldPath = common::cwd();
+  this->dataPtr->doCleanup = _cleanup;
+
+  auto tempPath = common::tempDirectoryPath();
+  if (!_subDir.empty())
+  {
+    tempPath = common::joinPaths(tempPath, _subDir);
+  }
+  this->dataPtr->path = common::createTempDirectory(_prefix, tempPath);
+  if (!this->dataPtr->path.empty())
+  {
+    this->dataPtr->isValid = true;
+    common::chdir(this->dataPtr->path);
```

Comment:
Can you document that the working directory will be changed at construction and destruction?

Reply:
ee7cae2